src/Akka.Streams.Kafka/Stages/CommitableConsumerStage.cs

```diff
@@ -0,0 +1,179 @@
+                }
+                else
+                {
+                    PullQueue();
+                }
+            });
+        }
+
+        public override void PreStart()
+        {
+            base.PreStart();
+
+            _consumer = _settings.CreateKafkaConsumer();
+            _consumer.OnMessage += HandleOnMessage;
+            _consumer.OnConsumeError += HandleConsumeError;
+            _consumer.OnError += HandleOnError;
+            _consumer.OnPartitionsAssigned += HandleOnPartitionsAssigned;
+            _consumer.OnPartitionsRevoked += HandleOnPartitionsRevoked;
+
+            switch (_subscription)
+            {
+                case TopicSubscription ts:
+                    _consumer.Subscribe(ts.Topics);
+                    break;
+                case Assignment a:
+                    _consumer.Assign(a.TopicPartitions);
+                    break;
+                case AssignmentWithOffset awo:
+                    _consumer.Assign(awo.TopicPartitions);
+                    break;
+            }
+
+            _messagesReceived = GetAsyncCallback<Message<K, V>>(MessagesReceived);
+            _partitionsAssigned = GetAsyncCallback<IEnumerable<TopicPartition>>(PartitionsAssigned);
+            _partitionsRevoked = GetAsyncCallback<IEnumerable<TopicPartition>>(PartitionsRevoked);
+        }
+
+        public override void PostStop()
+        {
+            _consumer.OnMessage -= HandleOnMessage;
+            _consumer.OnConsumeError -= HandleConsumeError;
+            _consumer.OnError -= HandleOnError;
+            _consumer.OnPartitionsAssigned -= HandleOnPartitionsAssigned;
+            _consumer.OnPartitionsRevoked -= HandleOnPartitionsRevoked;
+
+            _consumer.Dispose();
+
+            base.PostStop();
+        }
+
+        //
+        // Consumer's events
+        //
+
+        private void HandleOnMessage(object sender, Message<K, V> message) => _messagesReceived.Invoke(message);
+
+        // TODO: how I should react?
+        private void HandleConsumeError(object sender, Message message) { }
+
+        private void HandleOnError(object sender, Error error)
+        {
+            if (error.Code == ErrorCode.Local_Transport)
+            {
+                FailStage(new Exception(error.Reason));
+            }
+
+            // TODO: what else errors to handle?
+            Log.Error(error.Reason);
+        }
+
+        private void HandleOnPartitionsAssigned(object sender, List<TopicPartition> list)
+        {
+            _partitionsAssigned.Invoke(list);
+        }
+
+        private void HandleOnPartitionsRevoked(object sender, List<TopicPartition> list)
+        {
+            _partitionsRevoked.Invoke(list);
+        }
+
+        //
+        // Async callbacks
+        //
+
+        private void MessagesReceived(Message<K, V> message)
+        {
+            var consumer = _consumer;
+            var commitableOffset = new CommitableOffset(
+                () => consumer.CommitAsync(),
```

Comment:
Is it fine to pass Task here? Is it thread-safe?

Reply:
In terms of Akka it is since the task doesn't touch anything from the stage put I don't know what happens with the consumer if you poll/commit from different threads. Also why passing a func around ? Wouldn't it be easier to pass the consumer directly into the CommittableOffsset class and call _consumer.CommitAsync ? Another question, what happens if one calls Commit multiple times on the same message ? Should it be possible ?